code/game/objects/structures/safe.dm

```diff
@@ -211,96 +212,80 @@ GLOBAL_LIST_EMPTY(safes)
-		to_chat(user, "<span class='warning'>You're not able to operate the safe.</span>")
+/obj/structure/safe/proc/start_turning_knob(mob/user, canhear, ticks, left)
+	if(spinning || open)
 		return
 
-	var/canhear = FALSE
-	if(ishuman(user))
-		var/mob/living/carbon/human/H = user
-		if(H.can_hear() && H.is_in_hands(/obj/item/clothing/accessory/stethoscope))
-			canhear = TRUE
-
-	if(href_list["open"])
-		if(check_unlocked() || open || broken)
-			to_chat(user, "<span class='notice'>You [open ? "close" : "open"] [src].</span>")
-			open = !open
-			update_icon()
-		else
-			to_chat(user, "<span class='warning'>You can't open [src], as its lock is engaged!</span>")
-
-		. = TRUE
-
-	if(href_list["turnright"])
-		if(open)
-			return
-
-		if(broken)
-			to_chat(user, "<span class='warning'>The dial will not turn, as the mechanism is destroyed.</span>")
-			return
+	if(broken)
+		to_chat(user, "<span class='warning'>The dial will not turn, as the mechanism is destroyed.</span>")
+		return
 
-		var/ticks = text2num(href_list["turnright"])
-		for(var/i = 1 to ticks)
-			dial = WRAP(dial - 1, 0, 100)
+	spinning = TRUE
 
-			var/invalid_turn = current_tumbler_index % 2 == 0 || current_tumbler_index > number_of_tumblers
-			if(invalid_turn) // The moment you turn the wrong way or go too far, the tumblers reset
-				current_tumbler_index = 1
+	var/wrap_delta = left ? 1 : -1
+	for(var/i = 1 to ticks)
+		dial = WRAP(dial + wrap_delta, 0, 100)
 
-			if(!invalid_turn && dial == tumblers[current_tumbler_index])
-				notify_user(user, canhear, list("tink", "krink", "plink"), ticks, i)
-				current_tumbler_index++
-			else
-				notify_user(user, canhear, list("clack", "scrape", "clank"), ticks, i)
+		var/even_tumbler = current_tumbler_index % 2 == 0
+		var/invalid_turn = (left && !even_tumbler) || (!left && even_tumbler) || current_tumbler_index > number_of_tumblers
+		if(invalid_turn) // The moment you turn the wrong way or go too far, the tumblers reset
+			current_tumbler_index = 1
 
-			sleep(1)
-			check_unlocked(user, canhear)
-			SSnanoui.update_uis(src)
+		if(!invalid_turn && dial == tumblers[current_tumbler_index])
+			notify_user(user, canhear, list("tonk", "krunk", "plunk"), ticks, i)
+			current_tumbler_index++
+		else
+			notify_user(user, canhear, list("click", "chink", "clink"), ticks, i)
 
-		. = TRUE
+	addtimer(CALLBACK(src, .proc/finish_turning_knob, user, canhear), 1)
 
-	if(href_list["turnleft"])
-		if(open)
-			return
+/obj/structure/safe/proc/finish_turning_knob(mob/user, canhear)
+	check_unlocked(user, canhear)
+	SStgui.update_uis(src)
+	spinning = FALSE
 
-		if(broken)
-			to_chat(user, "<span class='warning'>The dial will not turn, as the mechanism is destroyed.</span>")
-			return
 
-		var/ticks = text2num(href_list["turnleft"])
-		for(var/i = 1 to ticks)
-			dial = WRAP(dial + 1, 0, 100)
+/obj/structure/safe/tgui_act(action, list/params)
+	var/mob/user = usr
+	if(!user.IsAdvancedToolUser() && !isobserver(user))
```

Comment:
I am questioning the need for the ` && !isobserver()` check, especially since the TGUI safeties will check the state. You are also missing those. Please add
```dm
if(..())
	return
```
To the very top of `tgui_act()`